File: scenarios/basic/cross-platform/qt/SilentLoadingState.cpp
#include <QApplication>
#include <QMainWindow>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QWidget>
#include <QTimer>
#include <QProgressBar>

class DataDashboard : public QMainWindow
{
    Q_OBJECT

public:
    DataDashboard(QWidget *parent = nullptr) : QMainWindow(parent)
    {
        setupUI();
        setupConnections();
    }

private slots:
    void refreshData()
    {
        startLoading();
    }
    
    void exportData()
    {
        startLoading();
    }
    
    void generateReport()
    {
        startLoading();
    }
    
    void updateLoading()
    {
        if (isLoading) {
            progress += 2;
            updateLoadingDisplay();
            
            if (progress >= 100) {
                completeLoading();
            }
        }
    }

private:
    void setupUI()
    {
        QWidget *centralWidget = new QWidget(this);
        setCentralWidget(centralWidget);
        
        QVBoxLayout *mainLayout = new QVBoxLayout(centralWidget);
        
        // Header
        QHBoxLayout *headerLayout = new QHBoxLayout();
        
        QLabel *titleLabel = new QLabel("Data Dashboard", this);
        titleLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 28px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    padding: 20px 0;"
            "}"
        );
        headerLayout->addWidget(titleLabel);
        
        headerLayout->addStretch();
        
        QPushButton *refreshBtn = new QPushButton("🔄 Refresh", this);
        refreshBtn->setStyleSheet(
            "QPushButton {"
            "    background: #007bff;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 6px;"
            "    padding: 10px 20px;"
            "    font-size: 16px;"
            "    font-weight: bold;"
            "    margin-right: 10px;"
            "}"
            "QPushButton:hover {"
            "    background: #0056b3;"
            "}"
        );
        connect(refreshBtn, &QPushButton::clicked, this, &DataDashboard::refreshData);
        headerLayout->addWidget(refreshBtn);
        
        QPushButton *exportBtn = new QPushButton("📊 Export", this);
        exportBtn->setStyleSheet(
            "QPushButton {"
            "    background: #28a745;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 6px;"
            "    padding: 10px 20px;"
            "    font-size: 16px;"
            "    font-weight: bold;"
            "    margin-right: 10px;"
            "}"
            "QPushButton:hover {"
            "    background: #218838;"
            "}"
        );
        connect(exportBtn, &QPushButton::clicked, this, &DataDashboard::exportData);
        headerLayout->addWidget(exportBtn);
        
        QPushButton *reportBtn = new QPushButton("📋 Report", this);
        reportBtn->setStyleSheet(
            "QPushButton {"
            "    background: #6c757d;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 6px;"
            "    padding: 10px 20px;"
            "    font-size: 16px;"
            "    font-weight: bold;"
            "}"
            "QPushButton:hover {"
            "    background: #5a6268;"
            "}"
        );
        connect(reportBtn, &QPushButton::clicked, this, &DataDashboard::generateReport);
        headerLayout->addWidget(reportBtn);
        
        mainLayout->addLayout(headerLayout);
        
        // Page header
        QHBoxLayout *pageHeaderLayout = new QHBoxLayout();
        
        QLabel *pageTitle = new QLabel("Project Overview", this);
        pageTitle->setStyleSheet(
            "QLabel {"
            "    font-size: 24px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    margin-bottom: 10px;"
            "}"
        );
        pageHeaderLayout->addWidget(pageTitle);
        
        pageHeaderLayout->addStretch();
        
        QLabel *pageSubtitle = new QLabel("Monitor and manage your projects", this);
        pageSubtitle->setStyleSheet(
            "QLabel {"
            "    font-size: 16px;"
            "    color: #666;"
            "    margin-bottom: 20px;"
            "}"
        );
        pageHeaderLayout->addWidget(pageSubtitle);
        
        mainLayout->addLayout(pageHeaderLayout);
        
        // SILENT LOADING STATE - Content spinner without screen reader announcement
        loadingOverlay = new QWidget(this);
        loadingOverlay->setStyleSheet(
            "QWidget {"
            "    background: rgba(248, 249, 250, 0.9);"
            "    border-radius: 12px;"
            "    padding: 40px;"
            "    margin: 20px 0;"
            "}"
        );
        loadingOverlay->setVisible(false);
        
        QVBoxLayout *loadingLayout = new QVBoxLayout(loadingOverlay);
        loadingLayout->setAlignment(Qt::AlignCenter);
        
        // SILENT LOADING STATE - Content spinner without screen reader announcement
        loadingSpinner = new QLabel("⏳", this);
        loadingSpinner->setStyleSheet(
            "QLabel {"
            "    font-size: 48px;"
            "    text-align: center;"
            "    margin-bottom: 20px;"
            "}"
        );
        loadingSpinner->setAlignment(Qt::AlignCenter);
        loadingLayout->addWidget(loadingSpinner);
        
        // SILENT LOADING STATE - Content spinner without screen reader announcement
        loadingText = new QLabel("Loading...", this);
        loadingText->setStyleSheet(
            "QLabel {"
            "    font-size: 18px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    text-align: center;"
            "    margin-bottom: 20px;"
            "}"
        );
        loadingText->setAlignment(Qt::AlignCenter);
        loadingLayout->addWidget(loadingText);
        
        // SILENT LOADING STATE - Content spinner without screen reader announcement
        progressBar = new QProgressBar(this);
        progressBar->setRange(0, 100);
        progressBar->setValue(0);
        progressBar->setStyleSheet(
            "QProgressBar {"
            "    border: 1px solid #ddd;"
            "    border-radius: 4px;"
            "    text-align: center;"
            "    font-weight: bold;"
            "    margin-bottom: 10px;"
            "}"
            "QProgressBar::chunk {"
            "    background: #007bff;"
            "    border-radius: 4px;"
            "}"
        );
        loadingLayout->addWidget(progressBar);
        
        // SILENT LOADING STATE - Content spinner without screen reader announcement
        progressText = new QLabel("0%", this);
        progressText->setStyleSheet(
            "QLabel {"
            "    font-size: 14px;"
            "    color: #666;"
            "    font-weight: bold;"
            "    text-align: center;"
            "}"
        );
        progressText->setAlignment(Qt::AlignCenter);
        loadingLayout->addWidget(progressText);
        
        mainLayout->addWidget(loadingOverlay);
        
        // Stats grid
        QHBoxLayout *statsLayout = new QHBoxLayout();
        
        QFrame *stat1 = createStatCard("📊", "12", "Total Projects");
        statsLayout->addWidget(stat1);
        
        QFrame *stat2 = createStatCard("✅", "8", "Completed");
        statsLayout->addWidget(stat2);
        
        QFrame *stat3 = createStatCard("⏳", "3", "In Progress");
        statsLayout->addWidget(stat3);
        
        QFrame *stat4 = createStatCard("⏸️", "1", "On Hold");
        statsLayout->addWidget(stat4);
        
        mainLayout->addLayout(statsLayout);
        
        // Data section
        QFrame *dataFrame = new QFrame(this);
        dataFrame->setStyleSheet(
            "QFrame {"
            "    background: white;"
            "    border: 1px solid #ddd;"
            "    border-radius: 8px;"
            "    padding: 25px;"
            "    margin: 20px 0;"
            "}"
        );
        
        QVBoxLayout *dataLayout = new QVBoxLayout(dataFrame);
        
        QHBoxLayout *dataHeaderLayout = new QHBoxLayout();
        
        QLabel *dataTitle = new QLabel("Recent Projects", this);
        dataTitle->setStyleSheet(
            "QLabel {"
            "    font-size: 20px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    margin-bottom: 20px;"
            "}"
        );
        dataHeaderLayout->addWidget(dataTitle);
        
        dataHeaderLayout->addStretch();
        
        QPushButton *filterBtn = new QPushButton("🔍 Filter", this);
        filterBtn->setStyleSheet(
            "QPushButton {"
            "    background: #f8f9fa;"
            "    color: #333;"
            "    border: 1px solid #ddd;"
            "    border-radius: 4px;"
            "    padding: 8px 16px;"
            "    font-size: 14px;"
            "    margin-right: 10px;"
            "}"
            "QPushButton:hover {"
            "    background: #e9ecef;"
            "}"
        );
        dataHeaderLayout->addWidget(filterBtn);
        
        QPushButton *sortBtn = new QPushButton("↕️ Sort", this);
        sortBtn->setStyleSheet(
            "QPushButton {"
            "    background: #f8f9fa;"
            "    color: #333;"
            "    border: 1px solid #ddd;"
            "    border-radius: 4px;"
            "    padding: 8px 16px;"
            "    font-size: 14px;"
            "}"
            "QPushButton:hover {"
            "    background: #e9ecef;"
            "}"
        );
        dataHeaderLayout->addWidget(sortBtn);
        
        dataLayout->addLayout(dataHeaderLayout);
        
        // Project list
        QLabel *projectList = new QLabel(
            "• Project Alpha - Active - 75% complete\n"
            "• Project Beta - Pending - 25% complete\n"
            "• Project Gamma - Completed - 100% complete\n"
            "• Project Delta - Active - 50% complete\n"
            "• Project Epsilon - On Hold - 10% complete",
            this
        );
        projectList->setStyleSheet(
            "QLabel {"
            "    font-size: 16px;"
            "    color: #333;"
            "    line-height: 1.6;"
            "    margin-top: 15px;"
            "}"
        );
        dataLayout->addWidget(projectList);
        
        mainLayout->addWidget(dataFrame);
        
        // Activity section
        QFrame *activityFrame = new QFrame(this);
        activityFrame->setStyleSheet(
            "QFrame {"
            "    background: white;"
            "    border: 1px solid #ddd;"
            "    border-radius: 8px;"
            "    padding: 25px;"
            "    margin: 20px 0;"
            "}"
        );
        
        QVBoxLayout *activityLayout = new QVBoxLayout(activityFrame);
        
        QLabel *activityTitle = new QLabel("Recent Activity", this);
        activityTitle->setStyleSheet(
            "QLabel {"
            "    font-size: 20px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    margin-bottom: 15px;"
            "}"
        );
        activityLayout->addWidget(activityTitle);
        
        QLabel *activityList = new QLabel(
            "🔄 Data refreshed - 2 minutes ago\n"
            "📊 Report generated - 15 minutes ago\n"
            "📤 Data exported - 1 hour ago\n"
            "✅ Project completed - 2 hours ago",
            this
        );
        activityList->setStyleSheet(
            "QLabel {"
            "    font-size: 16px;"
            "    color: #333;"
            "    line-height: 1.6;"
            "}"
        );
        activityLayout->addWidget(activityList);
        
        mainLayout->addWidget(activityFrame);
        
        // Status bar
        QHBoxLayout *statusLayout = new QHBoxLayout();
        
        QLabel *statusLabel = new QLabel("Ready", this);
        statusLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 14px;"
            "    color: #28a745;"
            "    font-weight: bold;"
            "    padding: 10px 0;"
            "}"
        );
        statusLayout->addWidget(statusLabel);
        
        statusLayout->addStretch();
        
        QLabel *lastUpdated = new QLabel("Last updated: " + QTime::currentTime().toString(), this);
        lastUpdated->setStyleSheet(
            "QLabel {"
            "    font-size: 14px;"
            "    color: #666;"
            "    padding: 10px 0;"
            "}"
        );
        statusLayout->addWidget(lastUpdated);
        
        mainLayout->addLayout(statusLayout);
    }
    
    QFrame* createStatCard(const QString &icon, const QString &value, const QString &label)
    {
        QFrame *card = new QFrame(this);
        card->setStyleSheet(
            "QFrame {"
            "    background: white;"
            "    border: 1px solid #ddd;"
            "    border-radius: 8px;"
            "    padding: 25px;"
            "    margin: 5px;"
            "}"
        );
        
        QVBoxLayout *cardLayout = new QVBoxLayout(card);
        cardLayout->setSpacing(15);
        
        QLabel *iconLabel = new QLabel(icon, this);
        iconLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 36px;"
            "    text-align: center;"
            "    margin-bottom: 10px;"
            "}"
        );
        iconLabel->setAlignment(Qt::AlignCenter);
        cardLayout->addWidget(iconLabel);
        
        QLabel *valueLabel = new QLabel(value, this);
        valueLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 28px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    text-align: center;"
            "    margin-bottom: 5px;"
            "}"
        );
        valueLabel->setAlignment(Qt::AlignCenter);
        cardLayout->addWidget(valueLabel);
        
        QLabel *labelLabel = new QLabel(label, this);
        labelLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 14px;"
            "    color: #666;"
            "    text-align: center;"
            "    text-transform: uppercase;"
            "}"
        );
        labelLabel->setAlignment(Qt::AlignCenter);
        cardLayout->addWidget(labelLabel);
        
        return card;
    }
    
    void setupConnections()
    {
        // Timer for loading animation
        loadingTimer = new QTimer(this);
        connect(loadingTimer, &QTimer::timeout, this, &DataDashboard::updateLoading);
    }
    
    void startLoading()
    {
        isLoading = true;
        progress = 0;
        loadingOverlay->setVisible(true);
        loadingText->setText("Loading data...");
        loadingTimer->start(100); // Update every 100ms
        updateLoadingDisplay();
    }
    
    void completeLoading()
    {
        isLoading = false;
        loadingTimer->stop();
        loadingOverlay->setVisible(false);
        progress = 0;
    }
    
    void updateLoadingDisplay()
    {
        if (isLoading) {
            progressBar->setValue(progress);
            progressText->setText(QString::number(progress) + "%");
            
            // Update loading message based on progress
            if (progress < 30) {
                loadingText->setText("Connecting to server...");
            } else if (progress < 60) {
                loadingText->setText("Fetching data...");
            } else if (progress < 90) {
                loadingText->setText("Processing results...");
            } else {
                loadingText->setText("Finalizing...");
            }
        }
    }

private:
    QWidget *loadingOverlay;
    QLabel *loadingSpinner;
    QLabel *loadingText;
    QProgressBar *progressBar;
    QLabel *progressText;
    QTimer *loadingTimer;
    bool isLoading = false;
    int progress = 0;
};

int main(int argc, char *argv[])
{
    QApplication app(argc, argv);
    
    DataDashboard app;
    app.setWindowTitle("Data Dashboard - Silent Loading");
    app.resize(1000, 700);
    app.show();
    
    return app.exec();
}

#include "SilentLoadingState.moc"

File: scenarios/basic/desktop/linux-qt/SilentLoadingState.py
import sys
from PyQt5.QtWidgets import (QApplication, QMainWindow, QVBoxLayout, QHBoxLayout, 
                             QWidget, QLabel, QPushButton, QFrame, QTableWidget, 
                             QTableWidgetItem, QHeaderView, QProgressBar, QSpinner)
from PyQt5.QtCore import Qt, QTimer
from PyQt5.QtGui import QFont

class SilentLoadingState(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("Data Dashboard")
        self.setGeometry(100, 100, 1000, 700)
        
        # Central widget
        central_widget = QWidget()
        self.setCentralWidget(central_widget)
        
        # Main layout
        main_layout = QVBoxLayout(central_widget)
        main_layout.setSpacing(0)
        
        # Toolbar
        toolbar_layout = QHBoxLayout()
        toolbar_layout.setSpacing(10)
        toolbar_layout.setContentsMargins(10, 10, 10, 10)
        toolbar_layout.setAlignment(Qt.AlignCenter)
        
        # Action buttons
        refresh_button = QPushButton("Refresh Data")
        refresh_button.setStyleSheet("QPushButton { background-color: #007bff; color: white; font-weight: bold; }")
        refresh_button.setFixedSize(120, 40)
        refresh_button.clicked.connect(self.on_refresh_clicked)
        toolbar_layout.addWidget(refresh_button)
        
        export_button = QPushButton("Export Data")
        export_button.setFixedSize(120, 40)
        export_button.clicked.connect(self.on_export_clicked)
        toolbar_layout.addWidget(export_button)
        
        generate_report_button = QPushButton("Generate Report")
        generate_report_button.setFixedSize(120, 40)
        generate_report_button.clicked.connect(self.on_generate_report_clicked)
        toolbar_layout.addWidget(generate_report_button)
        
        main_layout.addLayout(toolbar_layout)
        
        # Loading view - SILENT LOADING STATE
        self.loading_frame = QFrame()
        self.loading_frame.setFrameStyle(QFrame.Box)
        self.loading_frame.setFixedHeight(100)
        self.loading_frame.setStyleSheet("QFrame { background-color: #f8f9fa; }")
        self.loading_frame.setVisible(False)
        
        loading_layout = QVBoxLayout(self.loading_frame)
        loading_layout.setSpacing(15)
        loading_layout.setContentsMargins(20, 20, 20, 20)
        loading_layout.setAlignment(Qt.AlignCenter)
        
        # SILENT LOADING STATE - Content spinner without screen reader announcement
        self.progress_spinner = QLabel("⏳")
        self.progress_spinner.setAlignment(Qt.AlignCenter)
        self.progress_spinner.setStyleSheet("font-size: 40px;")
        loading_layout.addWidget(self.progress_spinner)
        
        loading_label = QLabel("Loading...")
        loading_font = QFont()
        loading_font.setPointSize(16)
        loading_font.setBold(True)
        loading_label.setFont(loading_font)
        loading_label.setAlignment(Qt.AlignCenter)
        loading_layout.addWidget(loading_label)
        
        # SILENT LOADING STATE - Content spinner without screen reader announcement
        self.progress_bar = QProgressBar()
        self.progress_bar.setFixedWidth(300)
        self.progress_bar.setValue(0)
        loading_layout.addWidget(self.progress_bar)
        
        progress_label = QLabel("0%")
        progress_label.setAlignment(Qt.AlignCenter)
        loading_layout.addWidget(progress_label)
        
        main_layout.addWidget(self.loading_frame)
        
        # Main content
        content_widget = QWidget()
        content_layout = QVBoxLayout(content_widget)
        content_layout.setSpacing(20)
        content_layout.setContentsMargins(20, 20, 20, 20)
        
        # Data summary
        summary_label = QLabel("Data Summary")
        summary_font = QFont()
        summary_font.setPointSize(18)
        summary_font.setBold(True)
        summary_label.setFont(summary_font)
        content_layout.addWidget(summary_label)
        
        summary_grid = QHBoxLayout()
        summary_grid.setSpacing(20)
        
        summary_items = [
            ("📊", "Total Records", "1,234", "blue"),
            ("✅", "Completed", "987", "green"),
            ("⏳", "Pending", "247", "orange")
        ]
        
        for icon, title, value, color in summary_items:
            summary_frame = self.create_summary_card(icon, title, value, color)
            summary_grid.addWidget(summary_frame)
        
        content_layout.addLayout(summary_grid)
        
        # Recent activity
        activity_label = QLabel("Recent Activity")
        activity_label.setFont(summary_font)
        content_layout.addWidget(activity_label)
        
        activity_frame = QFrame()
        activity_frame.setFrameStyle(QFrame.Box)
        activity_frame.setStyleSheet("QFrame { background-color: #f8f9fa; }")
        
        activity_layout = QVBoxLayout(activity_frame)
        activity_layout.setSpacing(10)
        activity_layout.setContentsMargins(15, 15, 15, 15)
        
        activities = [
            ("🟢", "Data refreshed", "2 minutes ago"),
            ("📊", "Report generated", "15 minutes ago"),
            ("📤", "Data exported", "1 hour ago")
        ]
        
        for icon, title, time in activities:
            activity_row = QHBoxLayout()
            
            icon_label = QLabel(icon)
            activity_row.addWidget(icon_label)
            
            title_label = QLabel(title)
            title_font = QFont()
            title_font.setBold(True)
            title_label.setFont(title_font)
            activity_row.addWidget(title_label)
            
            time_label = QLabel(time)
            time_label.setAlignment(Qt.AlignRight)
            activity_row.addWidget(time_label)
            
            activity_layout.addLayout(activity_row)
        
        content_layout.addWidget(activity_frame)
        
        # Data table
        table_label = QLabel("Data Records")
        table_label.setFont(summary_font)
        content_layout.addWidget(table_label)
        
        # Create table widget for data
        self.table_widget = QTableWidget()
        self.table_widget.setRowCount(5)
        self.table_widget.setColumnCount(4)
        self.table_widget.setHorizontalHeaderLabels(["ID", "Name", "Status", "Date"])
        self.table_widget.setFixedHeight(200)
        
        # Load sample data
        self.load_data()
        
        content_layout.addWidget(self.table_widget)
        
        main_layout.addWidget(content_widget)
        
        # State
        self.is_loading = False
        self.loading_progress = 0.0
        self.progress_timer = QTimer()
        self.progress_timer.timeout.connect(self.update_progress)
        
    def create_summary_card(self, icon, title, value, color):
        frame = QFrame()
        frame.setFrameStyle(QFrame.Box)
        frame.setFixedSize(200, 120)
        frame.setStyleSheet("QFrame { background-color: white; border: 1px solid #ddd; }")
        
        layout = QVBoxLayout(frame)
        layout.setSpacing(8)
        layout.setContentsMargins(15, 15, 15, 15)
        
        icon_label = QLabel(icon)
        icon_label.setAlignment(Qt.AlignCenter)
        icon_label.setStyleSheet("font-size: 36px;")
        layout.addWidget(icon_label)
        
        title_label = QLabel(title)
        title_font = QFont()
        title_font.setBold(True)
        title_label.setFont(title_font)
        title_label.setAlignment(Qt.AlignCenter)
        layout.addWidget(title_label)
        
        value_label = QLabel(value)
        value_font = QFont()
        value_font.setPointSize(16)
        value_font.setBold(True)
        value_label.setFont(value_font)
        value_label.setAlignment(Qt.AlignCenter)
        value_label.setStyleSheet(f"color: {color};")
        layout.addWidget(value_label)
        
        return frame
    
    def load_data(self):
        data = [
            ("1", "Record 1", "Active", "2024-01-01"),
            ("2", "Record 2", "Pending", "2024-01-02"),
            ("3", "Record 3", "Active", "2024-01-03"),
            ("4", "Record 4", "Pending", "2024-01-04"),
            ("5", "Record 5", "Active", "2024-01-05")
        ]
        
        for i, record in enumerate(data):
            for j, item in enumerate(record):
                self.table_widget.setItem(i, j, QTableWidgetItem(item))
    
    def start_loading(self):
        self.is_loading = True
        self.loading_frame.setVisible(True)
        self.loading_progress = 0.0
        
        # SILENT LOADING STATE - Content spinner without screen reader announcement
        self.progress_bar.setValue(0)
        
        # Start progress timer
        self.progress_timer.start(100)  # Update every 100ms
    
    def update_progress(self):
        if self.is_loading and self.loading_progress < 100:
            self.loading_progress += 2
            self.progress_bar.setValue(int(self.loading_progress))
            
            # Update progress label
            progress_label = self.loading_frame.findChild(QLabel)
            if progress_label and progress_label.text() != "Loading...":
                progress_label.setText(f"{int(self.loading_progress)}%")
        else:
            self.stop_loading()
    
    def stop_loading(self):
        self.is_loading = False
        self.loading_frame.setVisible(False)
        self.progress_timer.stop()
    
    def on_refresh_clicked(self):
        self.start_loading()
    
    def on_export_clicked(self):
        self.start_loading()
    
    def on_generate_report_clicked(self):
        self.start_loading()

def main():
    app = QApplication(sys.argv)
    window = SilentLoadingState()
    window.show()
    sys.exit(app.exec_())

if __name__ == "__main__":
    main()
